gms/gossiper.cc

```diff
@@ -731,33 +739,69 @@ future<> gossiper::do_status_check() {
     }
 }
 
-gossiper::endpoint_permit::endpoint_permit(endpoint_locks_map::entry_ptr&& ptr, semaphore_units<>&& units, inet_address addr, std::string caller) noexcept
-    : _permit(std::make_unique<permit>(std::move(ptr), std::move(units)))
+gossiper::endpoint_permit::endpoint_permit(endpoint_locks_map::entry_ptr&& ptr, inet_address addr, std::string caller) noexcept
+    : _ptr(std::move(ptr))
+    , _permit_id(_ptr->pid)
     , _addr(std::move(addr))
     , _caller(std::move(caller))
 {
-    logger.debug("{}: lock_endpoint {}: acquired", _caller, _addr);
+    ++_ptr->holders;
+    logger.debug("{}: lock_endpoint {}: acquired: permit_id={} holders={}", _caller, _addr, _permit_id, _ptr->holders);
 }
 
+gossiper::endpoint_permit::endpoint_permit(endpoint_permit&& o) noexcept
+    : _ptr(std::exchange(o._ptr, nullptr))
+    , _permit_id(std::exchange(o._permit_id, null_permit_id))
+    , _addr(std::exchange(o._addr, inet_address{}))
+    , _caller(std::move(o._caller))
+{}
+
 gossiper::endpoint_permit::~endpoint_permit() {
     release();
 }
 
 bool gossiper::endpoint_permit::release() noexcept {
-    if (auto permit = std::move(_permit)) {
-        logger.debug("{}: lock_endpoint {}: released", _caller, _addr);
-        return true;
+    if (auto ptr = std::exchange(_ptr, nullptr)) {
+        assert(ptr->pid == _permit_id);
+        logger.debug("{}: lock_endpoint {}: released: permit_id={} holders={}", _caller, _addr, _permit_id, ptr->holders);
+        if (!--ptr->holders) {
+            logger.debug("{}: lock_endpoint {}: released: permit_id={}", _caller, _addr, _permit_id);
+            ptr->units.return_all();
+            ptr->pid = null_permit_id;
+            _permit_id = null_permit_id;
+            return true;
+        }
     }
     return false;
 }
 
-future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep, seastar::compat::source_location l) {
+gossiper::endpoint_lock_entry::endpoint_lock_entry() noexcept
+    : sem(1)
+    , pid(permit_id::create_null_id())
+{}
+
+future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep, permit_id pid, seastar::compat::source_location l) {
     assert(this_shard_id() == 0);
     std::string caller = l.function_name();
-    logger.debug("{}: lock_endpoint {}: waiting", caller, ep);
-    auto eptr = co_await _endpoint_locks.get_or_load(ep, [] (const inet_address& ep) { return semaphore(1); });
-    auto units = co_await get_units(*eptr, 1);
-    co_return endpoint_permit(std::move(eptr), std::move(units), std::move(ep), std::move(caller));
+    auto eptr = co_await _endpoint_locks.get_or_load(ep, [] (const inet_address& ep) { return endpoint_lock_entry(); });
+    if (pid) {
```

Comment:
If the problem we're trying to solve is reentrancy, we don't need to pass permits. You can pass a flag `lock = false`, which skips the locking. The caller must hold the lock around the call. That would save a lot of complexity. 

Or you could split the API into two sets, external, which always lock, and internal, which doesn't, and can be called from listeners.

In general, we should avoid reentrancy, as that makes it hard to reason about gossiper state. Can we avoid calling gossiper from listeners?

Reply:
Later I've noticed that you're solving more than.